hash.c

```diff
@@ -6881,93 +6881,161 @@ static const rb_data_type_t env_data_type = {
- *    h[:nosuch] # => "Default value for nosuch"
+ *    h = Hash.new(false) # => {}
+ *    h[:foo]             # => false
  *
- *  Note that in the example above no entry for key +:nosuch+ is created:
+ *  Note: although the value of #default may be any object,
+ *  it may not be a good idea to use a mutable object:
  *
- *    h.include?(:nosuch) # => false
+ *    h = Hash.new([0, 1, 2])
+ *    h[:baz] # => [0, 1, 2]
+ *    h[:baz].reverse!
+ *    h[:baz] # => [2, 1, 0]
  *
- *  However, the proc itself can add a new entry:
+ *  ===== Per-Key Defaults
+ *
+ *  You can define a per-key default for a hash;
+ *  that is, a Proc that will return a value based on the key itself:
+ *
+ *  - The value of #default_proc must be a proc.
+ *  - The value of #default is +nil+ (always true when #default_proc is a proc).
+ *
+ *  You can set the default proc with method #default_proc=:
+ *
+ *    h = {foo: 0, bar: 1}
+ *    h.default_proc = proc {|hash, key| key.to_s.reverse }
+ *    h[:baz] # => "zab"
+ *
+ *  Or when the hash is created with Hash.new and a block:
+ *
+ *    h = Hash.new {|hash, key| key.to_s.upcase }
+ *    h[:baz] # => "BAZ"
+ *
+ *  In either case, the proc should accept two arguments:
+ *
+ *  - +hash+: the hash +self+.
+ *  - +key+: the key to be handled.
+ *
+ *  Note that the proc can modify +self+:
+ *
+ *    h = {foo: 0, bar: 1} # => {:foo=>0, :bar=>1}
+ *    h.default_proc = proc {|hash, key| hash.transform_keys! {|k| k.to_s } }
+ *    h[:baz]              # => {"foo"=>0, "bar"=>1}
+ *
+ *  Modifying +self+ in this way is not thread-safe;
+ *  multiple threads can concurrently call into the default proc
+ *  for the same key.
+ *
+ *  ==== \Method Default
+ *
+ *  For two methods, you can specify a default value for a not-found key
+ *  that has effect only for a single method call
+ *  (and not for any subsequent calls):
+ *
+ *  - #fetch(key).
+ *  - #fetch_values(*keys).
+ *
+ *  For method #fetch, you can specify an any-key default:
+ *
+ *    h = {foo: 0, bar: 1}
+ *    h.fetch(:nosuch, false) # => false
+ *    h.fetch(:nosuch)        # Raises KeyError.
  *
- *    synonyms = Hash.new { |hash, key| hash[key] = [] }
- *    synonyms.include?(:hello) # => false
- *    synonyms[:hello] << :hi # => [:hi]
- *    synonyms[:world] << :universe # => [:universe]
- *    synonyms.keys # => [:hello, :world]
+ *  For either method, you can specify a per-key default via a block:
  *
- *  Note that setting the default proc will clear the default value and vice versa.
+ *    h.fetch(:nosuch) {|key| "No key #{key}"}
+ *    # => "No key nosuch"
+ *    h.fetch(:nosuch)        # Raises KeyError.
  *
- *  Be aware that a default proc that modifies the hash is not thread-safe in the
- *  sense that multiple threads can call into the default proc concurrently for the
- *  same key.
+ *    h.fetch_values(:bar, :nosuch, :foo) {|key| "No key #{key}"}
+ *    # => [1, "No key nosuch", 0]
+ *    h.fetch_values(:bar, :nosuch, :foo) # Raises KeyError.
```

Comment:
IMO these sections have too much details about the methods themselves, which should belong in the method documentation rather than the class documentation. Class documentation should be about high level information, rather than the details about specific methods.

Reply:
New commit abbreviates the modified section, but does not (yet) make good the links in the 11 cited methods' doc; so still a draft.